src/server/session.ts

```diff
@@ -452,52 +425,201 @@ namespace ts.server {
         }
     }
 
-    type CombineProjectOutputCallback<TLocation extends DocumentPosition | undefined> = (
-        project: Project,
-        location: TLocation,
-        getMappedLocation: (project: Project, location: DocumentPosition) => DocumentPosition | undefined,
-    ) => void;
+    interface PerProjectResults<TResult> {
+        // The projects that were searched in the order in which they were searched.
+        // (The order may be slightly fudged to prioritize "authoritative" projects.)
+        projects: readonly Project[];
+        // The results for each project.
+        // May not have an entry for every member of `projects`.
+        resultsMap: ESMap<Project, readonly TResult[]>;
+    }
 
-    function combineProjectOutputWorker<TLocation extends DocumentPosition | undefined>(
+    /**
+     * @param projects Projects initially known to contain {@link initialPosition}
+     * @param defaultProject The default project containing {@link initialPosition}
+     * @param initialPosition Where the search operation was triggered
+     * @param getResultsForPosition This is where you plug in `findReferences`, `renameLocation`, etc
+     * @param forPositionInResult Given an item returned by {@link getResultsForPosition} enumerate the positions referred to by that result
+     * @returns In the common case where there's only one project, returns an array of results from {@link getResultsForPosition}.
+     * If multiple projects were searched - even if they didn't return results - the result will be a {@link PerProjectResults}.
+     */
+    function getPerProjectReferences<TResult>(

```

Comment:
This is the meat of the FAR logic.  I'd read it as standalone code and not try to map back to the old implementation.  If the comments aren't sufficient, I'd like to improve them.